crates/ruma-events/src/room/join_rules.rs

```diff
@@ -25,57 +24,53 @@ pub type JoinRulesEvent = StateEvent<JoinRulesEventContent>;
 pub struct JoinRulesEventContent {
     /// The type of rules used for users wishing to join this room.
     #[ruma_event(skip_redaction)]
+    #[serde(flatten)]
     pub join_rule: JoinRule,
-
-    /// Allow rules used for the `restricted` join rule.
-    #[cfg(feature = "unstable-pre-spec")]
-    #[serde(default)]
-    #[ruma_event(skip_redaction)]
-    pub allow: Vec<AllowRule>,
 }
 
 impl JoinRulesEventContent {
     /// Creates a new `JoinRulesEventContent` with the given rule.
     pub fn new(join_rule: JoinRule) -> Self {
-        Self {
-            join_rule,
-            #[cfg(feature = "unstable-pre-spec")]
-            allow: Vec::new(),
-        }
+        Self { join_rule }
     }
 
     /// Creates a new `JoinRulesEventContent` with the restricted rule and the given set of allow
     /// rules.
     #[cfg(feature = "unstable-pre-spec")]
     pub fn restricted(allow: Vec<AllowRule>) -> Self {
-        Self { join_rule: JoinRule::Restricted, allow }
+        Self { join_rule: JoinRule::Restricted(Restricted::new(allow)) }
     }
 }
 
 /// The rule used for users wishing to join this room.
 ///
 /// This type can hold an arbitrary string. To check for formats that are not available as a
 /// documented variant here, use its string representation, obtained through `.as_str()`.
-#[derive(Clone, Debug, PartialEq, Eq, StringEnum)]
-#[ruma_enum(rename_all = "lowercase")]
-#[non_exhaustive]
+#[derive(Clone, Debug, PartialEq, Eq, Serialize, Deserialize)]
```

Comment:
Unfortunately, the derived `Deserialize` impl won't work right for `_Custom`, i.e. it will expect `"join_rule": "_Custom"` rather than matching any value not covered by one of the other variants. This will require a manual `Deserialize` implementation.